Beam/Include/Beam/Threading/Sync.hpp

```diff
@@ -52,38 +130,53 @@ namespace Threading {
       */
       Sync(Sync&& sync);
 
-      //! Returns a reference of the value.
-      T& AcquireReference();
-
       //! Returns a copy of the value.
       T Acquire() const;
 
+      //! Acquire Sync's value in a synchronized manner.
+      /*!
+        \param f The action to perform on the value.
+      */
       template<typename F>
-      decltype(auto) With(F f) {
-        boost::unique_lock<Mutex> lock(m_mutex);
-        m_lock = &lock;
-        return f(m_value);
+      decltype(auto) With(const F& f) {
+        if constexpr(boost::hof::is_invocable<F, Value>()) {
+          return static_cast<const Sync*>(this)->With(f);
+        } else {
+          auto lock = WriteLock(m_mutex);
+          m_lock = LockProxy(&lock);
+          return f(m_value);
+        }
       }
 
+      //! Acquires Sync's value in a synchronized manner.
+      /*!
+        \param f The action to perform on the value.
+      */
       template<typename F>
-      decltype(auto) With(F f) const {
-        boost::unique_lock<Mutex> lock(m_mutex);
-        m_lock = &lock;
+      decltype(auto) With(const F& f) const {
+        auto lock = ReadLock(m_mutex);
+        m_lock = LockProxy(&lock);
         return f(m_value);
       }
 
+      //! Atomically acquires Sync's value along with another Sync's value.
+      /*!
+        \param s2 The other Sync.
+        \param f The action to perform on the values.
+      */
       template<typename S2, typename M2, typename F>
-      decltype(auto) With(Sync<S2, M2>& s2, F f) {
-        boost::unique_lock<Mutex> lock1(m_mutex, boost::defer_lock);
-        boost::unique_lock<M2> lock2(s2.m_mutex, boost::defer_lock);
+      decltype(auto) With(Sync<S2, M2>& s2, const F& f) {
+        auto lock1 = WriteLock(m_mutex, boost::defer_lock);
```

Comment:
As discussed, I did not work on this function except for making it aligned with the new implementation.
It is definitely possible to acquire read locks when possible for either `Sync`, and to make it work for any number of Syncs as well.